moby/moby
# Add --chown flag support for ADD/COPY commands for Windows

This is addressing: https://github.com/moby/moby/issues/35507

While the issue specifies warning that chown is not supported on
Windows, this fix actually implements chown support on Windows. Built-in
accounts as well as accounts included in the SAM database of the container
are supported.

The following are valid examples:

```Dockerfile
ADD --chown=Guest . <some directory>
COPY --chown=Administrator . <some directory>
COPY --chown=Guests . <some directory>
COPY --chown=ContainerUser . <some directory>
COPY --chown=someuser . <some directory>
```

On Windows an owner is only granted the permission to read the security
descriptor and read/write the discretionary access control list. This
fix also grants read/write and execute permissions to the owner.

Signed-off-by: Salahuddin Khan <salah@docker.com>

**- What I did**
Added support for --chown on Windows which supports built-in accounts including ContainerAdministrator/ContatinerUser and accounts specific to the container. Examples of the following include <container>\Administrator, <container>\Guest and any specific users to the container.

**- How I did it**
This works by invoking another binary built from https://github.com/salah-khan/windows-container-utility within the container and extracts the SID information.

**- How to verify it**

The following Dockerfile sets both container specific and well known groups and uses ICACLS to verify that the user/group in question has the access requested.

```FROM microsoft/windowsservercore
COPY --chown=Administrator . test
COPY --chown=Guests . test2
RUN icacls \test
RUN icacls \test\*
RUN icacls \test2
RUN icacls \test2\*
```
The following is the output of building with the Dockerfile:
```PS C:\testcode\chown> docker build -t test . --no-cache
Sending build context to Docker daemon  498.2kB
Step 1/7 : FROM microsoft/windowsservercore
 ---> ce27208ad678
Step 2/7 : COPY --chown=Administrator . test
 ---> e658217806cb
Step 3/7 : COPY --chown=Guests . test2
 ---> f6c1e7a3463d
Step 4/7 : RUN icacls \test
 ---> Running in 3dcba2ec54bd
\test BUILTIN\Administrators:(F)
      BUILTIN\Administrators:(OI)(CI)(IO)(F)
      NT AUTHORITY\SYSTEM:(F)
      NT AUTHORITY\SYSTEM:(OI)(CI)(IO)(F)
      3DCBA2EC54BD\Administrator:(R,W,D,WDAC)
      3DCBA2EC54BD\Administrator:(OI)(CI)(IO)(DE,Rc,WDAC,GR,GW)
      BUILTIN\Administrators:(I)(OI)(CI)(F)
      NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)
      3DCBA2EC54BD\Administrator:(I)(F)
      CREATOR OWNER:(I)(OI)(CI)(IO)(F)
      BUILTIN\Users:(I)(OI)(CI)(RX)
      BUILTIN\Users:(I)(CI)(AD)
      BUILTIN\Users:(I)(CI)(WD)

Successfully processed 1 files; Failed processing 0 files
Removing intermediate container 3dcba2ec54bd
 ---> 60b621288d53
Step 5/7 : RUN icacls \test\*
 ---> Running in 56295f6a34df
\test\Dockerfile BUILTIN\Administrators:(F)
                 BUILTIN\Administrators:(I)(F)
                 NT AUTHORITY\SYSTEM:(I)(F)
                 56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
                 BUILTIN\Users:(I)(RX)

\test\getsid.ps1 BUILTIN\Administrators:(F)
                 BUILTIN\Administrators:(I)(F)
                 NT AUTHORITY\SYSTEM:(I)(F)
                 56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
                 BUILTIN\Users:(I)(RX)

\test\OfflineSAMDBTest.exe BUILTIN\Administrators:(F)
                           BUILTIN\Administrators:(I)(F)
                           NT AUTHORITY\SYSTEM:(I)(F)
                           56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
                           BUILTIN\Users:(I)(RX)

\test\owner.exe BUILTIN\Administrators:(F)
                BUILTIN\Administrators:(I)(F)
                NT AUTHORITY\SYSTEM:(I)(F)
                56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
                BUILTIN\Users:(I)(RX)

\test\SAM BUILTIN\Administrators:(F)
          BUILTIN\Administrators:(I)(F)
          NT AUTHORITY\SYSTEM:(I)(F)
          56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
          BUILTIN\Users:(I)(RX)

\test\SAM.LOG1 BUILTIN\Administrators:(F)
               BUILTIN\Administrators:(I)(F)
               NT AUTHORITY\SYSTEM:(I)(F)
               56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
               BUILTIN\Users:(I)(RX)

\test\SAM.LOG2 BUILTIN\Administrators:(F)
               BUILTIN\Administrators:(I)(F)
               NT AUTHORITY\SYSTEM:(I)(F)
               56295F6A34DF\Administrator:(I)(R,W,D,WDAC)
               BUILTIN\Users:(I)(RX)

Successfully processed 7 files; Failed processing 0 files
Removing intermediate container 56295f6a34df
 ---> 4e6c9ad614d2
Step 6/7 : RUN icacls \test2
 ---> Running in dd923deb9a62
\test2 BUILTIN\Administrators:(F)
       BUILTIN\Administrators:(OI)(CI)(IO)(F)
       NT AUTHORITY\SYSTEM:(F)
       NT AUTHORITY\SYSTEM:(OI)(CI)(IO)(F)
       BUILTIN\Guests:(R,W,D,WDAC)
       BUILTIN\Guests:(OI)(CI)(IO)(DE,Rc,WDAC,GR,GW)
       BUILTIN\Administrators:(I)(OI)(CI)(F)
       NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)
       BUILTIN\Guests:(I)(F)
       CREATOR OWNER:(I)(OI)(CI)(IO)(F)
       BUILTIN\Users:(I)(OI)(CI)(RX)
       BUILTIN\Users:(I)(CI)(AD)
       BUILTIN\Users:(I)(CI)(WD)

Successfully processed 1 files; Failed processing 0 files
Removing intermediate container dd923deb9a62
 ---> 7eb8b1de9c87
Step 7/7 : RUN icacls \test2\*
 ---> Running in f140da0d1429
\test2\Dockerfile BUILTIN\Administrators:(F)
                  BUILTIN\Administrators:(I)(F)
                  NT AUTHORITY\SYSTEM:(I)(F)
                  BUILTIN\Guests:(I)(R,W,D,WDAC)
                  BUILTIN\Users:(I)(RX)

\test2\getsid.ps1 BUILTIN\Administrators:(F)
                  BUILTIN\Administrators:(I)(F)
                  NT AUTHORITY\SYSTEM:(I)(F)
                  BUILTIN\Guests:(I)(R,W,D,WDAC)
                  BUILTIN\Users:(I)(RX)

\test2\OfflineSAMDBTest.exe BUILTIN\Administrators:(F)
                            BUILTIN\Administrators:(I)(F)
                            NT AUTHORITY\SYSTEM:(I)(F)
                            BUILTIN\Guests:(I)(R,W,D,WDAC)
                            BUILTIN\Users:(I)(RX)

\test2\owner.exe BUILTIN\Administrators:(F)
                 BUILTIN\Administrators:(I)(F)
                 NT AUTHORITY\SYSTEM:(I)(F)
                 BUILTIN\Guests:(I)(R,W,D,WDAC)
                 BUILTIN\Users:(I)(RX)

\test2\SAM BUILTIN\Administrators:(F)
           BUILTIN\Administrators:(I)(F)
           NT AUTHORITY\SYSTEM:(I)(F)
           BUILTIN\Guests:(I)(R,W,D,WDAC)
           BUILTIN\Users:(I)(RX)

\test2\SAM.LOG1 BUILTIN\Administrators:(F)
                BUILTIN\Administrators:(I)(F)
                NT AUTHORITY\SYSTEM:(I)(F)
                BUILTIN\Guests:(I)(R,W,D,WDAC)
                BUILTIN\Users:(I)(RX)

\test2\SAM.LOG2 BUILTIN\Administrators:(F)
                BUILTIN\Administrators:(I)(F)
                NT AUTHORITY\SYSTEM:(I)(F)
                BUILTIN\Guests:(I)(R,W,D,WDAC)
                BUILTIN\Users:(I)(RX)

Successfully processed 7 files; Failed processing 0 files
Removing intermediate container f140da0d1429
 ---> 6ce8104c2ac9
Successfully built 6ce8104c2ac9
Successfully tagged test:latest
```
**- Description for the changelog**


Add --chown flag support for ADD/COPY commands for Windows

**- A picture of a cute animal (not mandatory but encouraged)**
